routers/web/web.go

```diff
@@ -1593,7 +1600,7 @@ func registerRoutes(m *web.Router) {
 	// end "/{username}/{reponame}": repo code
 
 	m.Group("/{username}/{reponame}", func() {
-		m.Get("/stars", repo.Stars)
+		m.Get("/stars", starsEnabled, repo.Stars)
 		m.Get("/watchers", repo.Watchers)
 		m.Get("/search", reqUnitCodeReader, repo.Search)
 		m.Post("/action/{action}", reqSignIn, repo.Action)
```

Comment:
IIRC here could also use middleware:

```
m.Post("/action/{action:star|unstar}", reqSignIn,  reqStarEnabled, repo.Action)
m.Post("/action/{action:watch|unwatch}", reqSignIn,  reqWatchEnabled, repo.Action)
m.Post("/action/{action:accept_transfer|reject_transfer}", reqSignIn, repo.Action)
```

Reply:
wouldn't that cause issues in case new actions were added? it would require the developer to know about these specific routes, unless there is a fallback for the action parameter.

an alternative would be to keep a single route and write a middleware that checks the action value.

Reply:
Nope. Actually the actions added to the "Action" function were not right. Each route handler should have clear behavior, but not do a lot of "switch" magic. You can see that the code in "Action" function is already a mess. No new action should/would be added anymore.

Now it is a good chance to make it clear again.